Impossible to update context of Virtual Router instances

### Description

After Virtual Router's instance(s) are created, any attempt of CONTEXT vector update causes panic.

```
panic: interface conversion: interface {} is nil, not []interface {}
```

This is particularly annoying because you can deploy fully working VR, but you can't reconfigure it, at the same time VR supports many features that can be dynamically enabled/disabled and reconfigured.


### Terraform and Provider version

```
Terraform v1.7.5
on linux_amd64
+ provider registry.terraform.io/opennebula/opennebula v1.4.0
```

### Affected resources and data sources

`opennebula_virtual_router_instance`

### Terraform configuration

```HCL
resource "opennebula_virtual_router_instance" "test" {
  name = "test"
  context = {
    ONEAPP_VNF_DNS_ENABLED = "NO"
  }
  virtual_router_id = opennebula_virtual_router.test.id
}
```


### Expected behavior

When I change `ONEAPP_VNF_DNS_ENABLED` from "NO" to "YES" and run `terraform apply` I want the VM re-contexted + no errors.

### Actual behavior

```
panic: interface conversion: interface {} is nil, not []interface {}
```

### Steps to Reproduce

1. Define the usual `opennebula_virtual_router`, `opennebula_virtual_router_instance_template`, and `opennebula_virtual_router_instance`, then apply.
2. Change any context variable of the `opennebula_virtual_router_instance`, then apply.

### Debug output

_No response_

### Panic output

_No response_

### Important factoids

The `resourceOpennebulaVirtualMachineRead` method seems to be throwing errors.

### References

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,10 @@ ENHANCEMENTS:
 
 * resources/opennebula_template: enable disk and nic update (#491)
 
+BUG FIXES:
+
+* resources/opennebula_virtual_router_instance: fix re-contextualization (#537)
+
 # 1.4.0 (January 22nd, 2024)
 
 FEATURES:

--- a/opennebula/resource_opennebula_virtual_machine.go
+++ b/opennebula/resource_opennebula_virtual_machine.go
@@ -605,16 +605,20 @@ func resourceOpennebulaVirtualMachineRead(ctx context.Context, d *schema.Resourc
 
 		var diags diag.Diagnostics
 
-		// read template ID from which the VM was created
-		templateID, _ := vmInfos.Template.GetInt("TEMPLATE_ID")
-		d.Set("template_id", templateID)
+		// NOTE: The template_id attribute is not defined for Virtual Router instances (VMs).
+		if _, ok := d.GetOk("template_id"); ok {
+			// read template ID from which the VM was created
+			templateID, _ := vmInfos.Template.GetInt("TEMPLATE_ID")
+			d.Set("template_id", templateID)
+
+			if _, ok := d.GetOk("template_nic"); !ok {
+				d.Set("template_nic", []interface{}{})
+			}
+		}
 
 		// add empty values for import
 		if _, ok := d.GetOk("template_disk"); !ok {
 			d.Set("template_disk", []interface{}{})
-		}
-		if _, ok := d.GetOk("template_nic"); !ok {
-			d.Set("template_nic", []interface{}{})
 		}
 		if _, ok := d.GetOk("template_tags"); !ok {
 			d.Set("template_tags", map[string]interface{}{})
@@ -632,14 +636,19 @@ func resourceOpennebulaVirtualMachineRead(ctx context.Context, d *schema.Resourc
 			})
 			return diags
 		}
-		err = flattenVMNIC(d, &vmInfos.Template)
-		if err != nil {
-			diags = append(diags, diag.Diagnostic{
-				Severity: diag.Error,
-				Summary:  "Failed to flatten NICs",
-				Detail:   fmt.Sprintf("virtual machine (ID: %s): %s", d.Id(), err),
-			})
-			return diags
+
+		// In case of Virtual Router instances (which are just VMs) there's never anything to "flatten",
+		// that's because NICs are attached with a help of dedicated resources.
+		if _, ok := d.GetOk("nic"); ok {
+			err = flattenVMNIC(d, &vmInfos.Template)
+			if err != nil {
+				diags = append(diags, diag.Diagnostic{
+					Severity: diag.Error,
+					Summary:  "Failed to flatten NICs",
+					Detail:   fmt.Sprintf("virtual machine (ID: %s): %s", d.Id(), err),
+				})
+				return diags
+			}
 		}
 
 		return nil

--- a/opennebula/resource_opennebula_virtual_router_test.go
+++ b/opennebula/resource_opennebula_virtual_router_test.go
@@ -91,6 +91,22 @@ func TestAccVirtualRouter(t *testing.T) {
 				),
 			},
 			{
+				Config: testAccVirtualRouterContextUpdate,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("opennebula_virtual_router_instance.test2", "context.update_test", "123"),
+					resource.TestCheckResourceAttrSet("opennebula_virtual_router.test", "uid"),
+					resource.TestCheckResourceAttrSet("opennebula_virtual_router.test", "gid"),
+					resource.TestCheckResourceAttrSet("opennebula_virtual_router.test", "uname"),
+					resource.TestCheckResourceAttrSet("opennebula_virtual_router.test", "gname"),
+					testAccCheckVirtualRouterPermissions(&shared.Permissions{
+						OwnerU: 1,
+						OwnerM: 1,
+						GroupU: 1,
+						OtherM: 1,
+					}, "testacc-vr"),
+				),
+			},
+			{
 				Config: testAccVirtualRouterAddNICs,
 				Check: resource.ComposeTestCheckFunc(
 					// Virtual router
@@ -352,6 +368,46 @@ resource "opennebula_virtual_router_instance" "test2" {
 	cpu = 0.1
 
 	virtual_router_id = opennebula_virtual_router.test.id
+}
+
+resource "opennebula_virtual_router" "test" {
+  name = "testacc-vr"
+  permissions = "642"
+  group = "oneadmin"
+
+  instance_template_id = opennebula_virtual_router_instance_template.test.id
+
+  tags = {
+    customer = "1"
+  }
+}
+`
+
+var testAccVirtualRouterContextUpdate = testAccVirtualRouterMachineTemplate + `
+
+resource "opennebula_virtual_router_instance" "test" {
+	name        = "testacc-vr-virtual-machine"
+	group       = "oneadmin"
+	permissions = "642"
+	memory = 128
+	cpu = 0.1
+
+	virtual_router_id = opennebula_virtual_router.test.id
+}
+
+
+resource "opennebula_virtual_router_instance" "test2" {
+	name        = "testacc-vr-virtual-machine-2"
+	group       = "oneadmin"
+	permissions = "642"
+	memory = 128
+	cpu = 0.1
+
+	virtual_router_id = opennebula_virtual_router.test.id
+
+	context = {
+	  update_test = "123"
+	}
 }
 
 resource "opennebula_virtual_router" "test" {

